lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,88 +3642,1015 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    return loaded;
+  }
+
+  SILValue materializeTangentDirect(
+      AdjointValue val, SILLocation loc) {
+    assert(val.getType().isObject());
+    LLVM_DEBUG(getADDebugStream() <<
+               "Materializing tangents for " << val << '\n');
+    assert(val.getKind() != AdjointValueKind::Aggregate
+           && "Tangent values cannot be aggregated in forward mode.");
+    switch (val.getKind()) {
+      case AdjointValueKind::Zero: {
+        auto zeroVal = emitZeroDirect(val.getSwiftType(), loc);
+        return zeroVal;
+      }
+      // Aggregate case here is necessary until we update 'AdjointValue' to be
+      // 'TangentValue'. An assertion is made above to make sure we
+      // never see an Aggregate in forward mode until this change is made.
+      case AdjointValueKind::Aggregate:
+      case AdjointValueKind::Concrete:
+        return val.getConcreteValue();
+    }
+  }
+
+  SILValue materializeTangent(AdjointValue val, SILLocation loc) {
+    if (val.isConcrete()) {
+      LLVM_DEBUG(getADDebugStream()
+                 << "Materializing tangent: Value is concrete.\n");
+      return val.getConcreteValue();
+    }
+    LLVM_DEBUG(getADDebugStream() << "Materializing tangent: Value is "
+                                     "non-concrete. Materializing directly.\n");
+    return materializeTangentDirect(val, loc);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Tangent buffer mapping
+  //--------------------------------------------------------------------------//
+
+  void setTangentBuffer(SILBasicBlock *origBB, SILValue originalBuffer,
+                        SILValue adjointBuffer) {
+    assert(originalBuffer->getType().isAddress());
+    auto insertion =
+        bufferMap.try_emplace({origBB, originalBuffer}, adjointBuffer);
+    assert(insertion.second); (void)insertion;
+  }
+
+  SILValue &getTangentBuffer(SILBasicBlock *origBB,
+                             SILValue originalBuffer) {
+    auto diffBuilder = getDifferentialBuilder();
+    assert(originalBuffer->getType().isAddress());
+    assert(originalBuffer->getFunction() == original);
+    auto insertion = bufferMap.try_emplace({origBB, originalBuffer},
+                                           SILValue());
+    if (!insertion.second) // not inserted
+      return insertion.first->getSecond();
+
+    // Set insertion point for local allocation builder: before the last local
+    // allocation, or at the start of the adjoint function's entry if no local
+    // allocations exist yet.
+    diffLocalAllocBuilder.setInsertionPoint(
+        getDifferential().getEntryBlock(),
+        getNextDifferentialLocalAllocationInsertionPoint());
+    // Allocate local buffer and initialize to zero.
+    auto *newBuf = diffLocalAllocBuilder.createAllocStack(
+        originalBuffer.getLoc(),
+        getRemappedTangentType(originalBuffer->getType()));
+    auto *access = diffLocalAllocBuilder.createBeginAccess(
```

Comment:
All `begin_access` and `end_access` pairs have been removed from AD-generated code in https://github.com/apple/swift/pull/26209/commits/ec158faca8b1f06af3768db00c4fea15505d4638. You should update this patch as well.